[GGB-145] 테스트 코드 관련 Util 코드 작성

JinTaekLim opened the pull request:
### **MemberRepositoryTest**
기존 -> FixtureMonkey
변경 -> RandomEntityPopulator

### **RandomEntityPopulator**

기존 모든 컬럼의 size와 null 여부를 Entity 파일에 정의하여 이를 읽고 값을 생성하는 형식으로 사용하였으나 
Jpa가 아닌 mybatis를 사용하기에 Entity 파일을 변경해도 테이블 구조가 변경되지 않아 SqlSessionFactory를 이용해 직접 이를 읽어오는 형식으로 구현 

랜덤 값은 무작위 값 ( 특수문자 , 한국어 , 영어 , 중국어 등등 ) 으로 지정되며, 컬럼별 제약 사항에 맞춰 알맞은 값을 반환하도록함

1. 랜덤 무작위 값 생성

`RandomEntityPopulator randomEntityPopulator = new RandomEntityPopulator(sqlSessionFactory)
randomEntityPopulator.getPopulatedEntity(Member.class,"MEMBERS");`

2. 랜덤 무작위 값 생성 , 특정 필드 값 임의 등록

`
RandomEntityPopulator randomEntityPopulator = new RandomEntityPopulator(sqlSessionFactory)
randomEntityPopulator
        .setValue("SnsType", snsType)
        .getPopulatedEntity(Member.class, "MEMBERS");
`



동작 순서

1. RandomEntityPopulator 생성시 sqlSessionFactory 주입하여 데이터베이스와 연결 

2. setValue 메서드를 사용하여 변경하고 싶은 컬럼명과 값을 지정 ( 선택사항 )
 값은 Map<String, Object> customValues에 저장되어 8번 과정에서 사용됨

4. getPopulatedEntity(Class<?> clazz, String tableName) 메서드로 엔티티 클래스와 대문자로 변환된 테이블 이름을 전달.

5. SqlSession을 열고, Connection을 통해 DatabaseMetaData를 조회

6. 전달 받은 엔티티 클래스를 이용해 새 객체 생성

7. 대문자로 변환한 엔티티의 변수값과 데이터베이스의 컬럼명을 비교해 제약 사항을 가져옴 ( size, null 여부 ) 
엔티티에 정의된 변수명과 데이터베이스의 컬럼명이 다를 경우 오류 발생. ( 대소문자 구분 X )

8. 제약 사항에 따라 랜덤값을 생성 후 반환 
customValues에 값이 존재할 경우, 해당 컬럼은 지정된 값으로 설정

### **CustomAssertions**

기존 테스트 코드에서 클래스가 동일한지 확인하기 위해서는 아래와 같은 코드가 필요했음
`assertEquals(member.getName() , testMember.getName());
assertEquals(member.getEmail() , testMember.getEmail());
...`

아래의 메소드를 사용하면 위와 같은 과정을 거치지 않고도 클래스의 내부의 모든 값을 비교해주며, 어떤 값이 저장되었고 비교되었는지 출력되어 확인할 수 있음

`customAssertEquals(Object obj1, Object obj2)`

### ++ 오류 사항 수정 ++

기존
-> getPopulatedEntity 사용시 전달 받은 엔티티 ( 클래스 ) 의 모든 변수를 읽고, 변수명과 DB의 컬럼명이 동일할 경우 DB의 제약 사항 ( size , null여부 ) 를 읽어 반환하였음.

변수명을 전달 받고, 해당 변수명과 동일한 컬럼이 있는지 매번 DB에 검색하는 방식이었기에 성능 저하와 DB 지나치게 의존적이었으며, 엔티티 ( 클래스 ) 에 정의되어있는 변수명 ( createId ) 과 DB 컬럼에 저장 되어있는 컬럼명 ( create_id ) 이 다를 경우 제약사항을 가져오지 못 하는 상황이 발생하였음. 
( 스네이크 케이스, 카멜 케이스 )

변경 
-> getPopulatedEntity 사용시 테이블의 모든 컬럼을 최초 한 번만 읽고, 컬럼명에 존재하는 '_' 특수문자를 제거한 후 ColumnDetail에 이름 , 크기 , null 여부를 저장하여 필요할 경우 꺼내 쓸 수 있도록 변경.






### **FixtureMonkey와 CustomValue의 반환 속도**

단순하게 size와 null여부만을 읽고 이에 맞는 랜덤 값을 생성하는 속도만을 비교해봤습니다.

특정 필드 값을 지정하거나 하는 등의 테스트도 하고 싶었는데,, fixtureMonkey 사용법을 도저히 모르겠어서 자세한 사항까지는 불가능했다는 점.. 양해부탁드립니다.


테스트 환경에 따라 조금씩 결과는 달라지겠지만, 약 100번 테스트까지는 유의미한 차이는 보이지 않지만 테스트가 많아지면 많아질 수록 fixtureMonkey 응답 속도가 빠른 것을 확인할 수 있습니다.


fixtureMonkey
10번 -> 1초192ms ( 첫 테스트 996 ms )
100번 -> 2초141ms ( 첫 테스트 1초 65 ms )
1000번 -> 6초185ms ( 첫 테스트 837 ms )


custom
10번 -> 1초29ms ( 첫 테스트 820 ms )
100번 -> 2초440ms ( 첫 테스트 1초 98 ms )
1000번 -> 9초479ms ( 첫 테스트 842 ms )
10000번 -> 1분7초 ( 첫 테스트 698ms ) 

custom2
10번 -> 536ms ( 첫 테스트 432ms )
100번 -> 1초129ms ( 첫 테스트 426ms )
1000번 -> 5초 699ms ( 첫 테스트 438ms )
10000번 -> 42초 60ms ( 첫 테스트 424ms )

** 9/15 + **
리펙토링 진행 전, 현재 어디서 성능 저하가 일어났는지 정확하게 확인하고 싶어서 각 메소드 별 실행 시간을 AOP로 체크하려 했습니다만, 
1. bean 등록되지 않은 클래스는 AOP 사용 불가
2. bean 등록된 클래스라도 내부 메소드는 일반적인 방법으로  AOP 사용 불가.

이러한 사항 때문에 임시로 RandomValue와 관련된 모든 파일에는 Component로 등록하고, RepositoryTest 파일 내에서 Autowired로 주입하는 형식으로 코드를 작성하였습니다.


**CustomAssertions - customAssertPresent 추가**

Optional 값의 존재 여부를 확인 , 값이 존재할 경우 해당 객체를 반환 / 존재 하지 않을 경우 AssertionError

JinTaekLim committed:
[GGB-145] RandomEntityPopulator , RandomValueGenerator　작성

( Entity의 값에 랜덤 값을 생성하고, 반환하는 코드 )

JinTaekLim committed:
[GGB-145] ColumnInfo 작성

JinTaekLim committed:
[GGB-145] MemberRepositoryTest
fixtureMonkey -> RandomEntityPopulator 변경

hyensukim reviewed (approved):
좋은 대안 매우 잘봤고, 끝까지 노력하셨던 집념에 박수를 쳐드리고 싶습니다. 고생 많으셨습니다.

한가지 궁금한 내용이 FixtureMonkey와 대안코드를 사용했을 때, 테스트 코드 성능(실행 시간)이 어떤지 비교 한번 해주셨으면, 좋겠습니다.

저도 이번에 테스트 코드를 진행하면서 알게된 내용인데, 테스트코드의 잘 작성된 원칙 중 하나가 Fast한 테스트를 진행하는 것도 있더라구요!

해당 내용은 추후 테스트 해보시고 공유해주시면 좋을것 같습니다 :)

JinTaekLim committed:
[GGB-145]
SQL 테이블 생성시 스네이크 케이스로 작성된 컬럼 Size 값을 읽지 못 하는 오류 수정

JinTaekLim committed:
[GGB-145]
class의 모든 값을 비교해주는 CustomAssertions 작성

JinTaekLim committed:
[GGB-145]
RandomValueGenerator , LocalDateTime 밀리초까지 생성되도록 변경

JinTaekLim committed:
[GGB-145]
CustomAssertions 클래스 내부 값 출력
System.out.print -> StringBuilder 변경

JinTaekLim committed:
[GGB-145]
MemberEntityPopulator 관련 코드 component 등록

JinTaekLim committed:
[GGB-145]
MemberEntityPopulator 내부 메소드 실행 시간 체크 및 출력

JinTaekLim committed:
[GGB-145]
8571c86 reset

JinTaekLim committed:
[GGB-145]
RandomEntityPopulator 생성과 동시에 tableName도 함께 전달하도록 변경 /
이전 -> getPopulatedEntity 호출마다 DB 조회
변경 -> RandomEntityPopulator 생성시에 단 한 번 DB 조회

JinTaekLim committed:
[GGB-145]
ExecutionTimeAspect 삭제

JinTaekLim committed:
[GGB-145]
MemberRepositoryTest IntegrationTest 상속 하도록 변경 및 Transactional 제거

JinTaekLim committed:
[GGB-145]
MemberRepositoryTest NotFoundMemberException 관련 코드 제거

JinTaekLim committed:
[GGB-145]
CustomAssertions , assertSameClass
[기존] throw BadRequestException
[변경] throw AssertionError

JinTaekLim committed:
[GGB-145]
CustomAssertions , customAssertEquals

[기존] catch 시 오류 메세지 출력
[변경] catch 시 throw AssertionError

JinTaekLim committed:
[GGB-145]
CustomAssertions , customAssertPresent (추가)

Optional 값의 존재 여부를 확인 , 값이 존재할 경우 해당 객체를 반환 / 존재 하지 않을 경우 AssertionError

JinTaekLim committed:
[GGB-145]
MemberRepositoryTest

[기존] Optional 값을 .orElse(null) 로 체크
[변경] customAssertPresent 를 사용해 체크
